### PR TITLE
[test] Ibex fault dump tests

### PR DESCRIPTION
**Pull request by HU90m**
The `rstmgr_cpu_info_test` now covers https://github.com/lowRISC/opentitan/issues/13219 and https://github.com/lowRISC/opentitan/issues/13218.

### CODE REVIEW

**Comment by tjaychen** [comment]
if possible, could you give this a try in `dvsim vcs` also? if you don't have things setup for that, we can help run also.
@jdonjdon

**Comment by HU90m** [comment]
> if possible, could you give this a try in `dvsim vcs` also? if you don't have things setup for that, we can help run also. @jdonjdon

It should pass in VCS already. `rstmgr_cpu_info_test` was broken in verilator, so I have exclusively been using VCS. (Xcelium was isn't running on my computer at the moment.)

**Comment by tjaychen** [comment]
sounds great, thank you.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 1]
```diff
@@ -29,171 +28,151 @@ OTTF_DEFINE_TEST_CONFIG();
...
-};
+// Declaring the labels used to calculate the expected current and next pc.
+extern const uint32_t _ottf_interrupt_vector, handler_exception;
 
-/**
- * Reserve expected cpu dump area in flash
- */
-OT_SECTION(".non_volatile_scratch")
-dif_rstmgr_cpu_info_dump_segment_t exp_dump[kCpuDumpSize];
+//
```
Comment (engdoreis) [inline sw/device/tests/rstmgr_cpu_info_test.c]: Should we have a comment here, if not just drop the `//`

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 2]
```diff
@@ -29,171 +28,151 @@ OTTF_DEFINE_TEST_CONFIG();
...
+        next_pc, observed.mnpc);
+  CHECK(last_data_addr == observed.mdaa,
+        "Last Data Access Addr: Expected 0x%x != Observed 0x%x", last_data_addr,
+        observed.mdaa);
+  CHECK(
+      mpec_lower <= observed.mpec && observed.mpec < mpec_upper,
+      "The Observed MPEC, 0x%x, was not in the expected range of [0x%x, 0x%x)",
+      observed.mpec, mpec_lower, mpec_upper);
+}
+static void check_previous_values(dif_rv_core_ibex_crash_dump_info_t *obs_dump,
```
Comment (engdoreis) [inline sw/device/tests/rstmgr_cpu_info_test.c:153]: Could you add a new line before this function for concistency?

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 3]
```diff
@@ -29,171 +28,151 @@ OTTF_DEFINE_TEST_CONFIG();
...
+        observed.mpec, mpec_lower, mpec_upper);
 }
 
 bool test_main(void) {
   dif_rstmgr_t rstmgr;
   dif_aon_timer_t aon_timer;
   dif_pwrmgr_t pwrmgr;
+  dif_rv_core_ibex_t ibex;
 
+  // Initialize Handles
```
Comment (engdoreis) [inline sw/device/tests/rstmgr_cpu_info_test.c:174]:
```suggestion
  // Initialize Handles.
```

File: sw/device/lib/dif/dif_rv_core_ibex.h [hunk 4]
```diff
@@ -201,7 +201,7 @@ typedef struct dif_rv_core_ibex_crash_dump_info {
   /**
    * The crash dump state for the current execution.
    */
-  dif_rv_core_ibex_crash_dump_state_t fault_stage;
+  dif_rv_core_ibex_crash_dump_state_t fault_state;
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_rv_core_ibex.h:204]: Nit. I would move this commit to the second position in the PR.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 5]
```diff
@@ -23,43 +23,78 @@ OTTF_DEFINE_TEST_CONFIG();
...
  *  This test creates a double fault by accessing a register with
  *  a non-existing address.
- *  After the double fault, the dut gets reset by the watch dog bite,
+ *  After a single fault and then a double fault,
+ *  the dut gets reset by the watch dog bite,
  *  and the test collects / checks the cpu_info from the rstmgr.
  */
 
-// non existing address
+// Unmapped Addresses
```
Comment (engdoreis) [inline sw/device/tests/rstmgr_cpu_info_test.c:31]:
```suggestion
// Unmapped Addresses.
```

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 6]
```diff
@@ -122,57 +181,95 @@ bool test_main(void) {
...
+      CHECK_DIF_OK(dif_rstmgr_cpu_info_set_enabled(&rstmgr, kDifToggleEnabled));
 
-    uint32_t bark_cycles = aon_timer_testutils_get_aon_cycles_from_us(100);
-    uint32_t bite_cycles = aon_timer_testutils_get_aon_cycles_from_us(100);
+      double_fault = false;
+      OT_ADDRESSABLE_LABEL(kSingleFaultAddrLower);
+      addr_val = *kIllegalAddr0;
+      OT_ADDRESSABLE_LABEL(kSingleFaultAddrUpper);
+      LOG_ERROR(
+          "This should be unreachable; a single fault should have occured.");
```
Comment (engdoreis) [inline sw/device/tests/rstmgr_cpu_info_test.c:196]: You'd better using `CHECK(false,"This should be unreachable; a single fault should have occured.")` here to make sure the test will not procede.

**Review by engdoreis — COMMENTED** [review]
Thanks for the PR @HU90m, I left some comments.

**Review by jdonjdon — APPROVED** [review]
Nice work!

**Comment by johngt** [comment]
@engdoreis can you please tag in development issues the following issues that will be closed out with this PR?
https://github.com/lowRISC/opentitan/issues/13218
https://github.com/lowRISC/opentitan/issues/13219

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 7]
```diff
@@ -24,176 +23,253 @@ OTTF_DEFINE_TEST_CONFIG();
...
+ *  After a single fault and then a double fault,
+ *  the dut gets reset by the watch dog bite,
  *  and the test collects / checks the cpu_info from the rstmgr.
  */
 
-// non existing address
-#define kIllegalAddr1 0x4041FFF0u
-#define kIllegalAddr2 0x40003618u
-#define kSkipComp 0x12345678u
+// Unmapped Addresses
```
Comment (luismarques) [inline sw/device/tests/rstmgr_cpu_info_test.c:31]: Nit: if you're going to change this comment anyway, just clarify that besides being invalid addresses they don't have any meaning, these are basically just random values, so they could also be `0xA0000000`, `0xA0000004`, `0xA0000008`, etc.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 8]
```diff
@@ -24,176 +23,253 @@ OTTF_DEFINE_TEST_CONFIG();
...
+static const uint8_t *kIllegalAddr0 = (uint8_t *)0xA0A0DEAFu;
+static const uint8_t *kIllegalAddr1 = (uint8_t *)0xA041FFF0u;
+static const uint8_t *kIllegalAddr2 = (uint8_t *)0xA0003618u;
```
Comment (luismarques) [inline sw/device/tests/rstmgr_cpu_info_test.c:34]: The `u` prefix is no longer needed. These could also be enums or preprocessor definitions, but maybe it's not worth changing that now. As discussed in person, they could also be `volatile` to avoid the assignment to `addr_val`, but since you have other uses for `addr_val` it's probably not worth changing that.

**Review by luismarques — APPROVED** [review]
Overall LGTM. Just some nitpicking issues left to correct and it can be merged.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 9]
```diff
@@ -24,176 +23,253 @@ OTTF_DEFINE_TEST_CONFIG();
...
+ *  the dut gets reset by the watch dog bite,
  *  and the test collects / checks the cpu_info from the rstmgr.
  */
 
-// non existing address
-#define kIllegalAddr1 0x4041FFF0u
-#define kIllegalAddr2 0x40003618u
-#define kSkipComp 0x12345678u
+// Unmapped Addresses
+static const uint8_t *kIllegalAddr0 = (uint8_t *)0xA0A0DEAFu;
```
Comment (alphan) [inline sw/device/tests/rstmgr_cpu_info_test.c:32]: These can fall in the translated address window (see [here](https://cs.opensource.google/opentitan/opentitan/+/master:hw/top_earlgrey/sw/autogen/top_earlgrey_memory.ld;drc=0ea3c9ccf15917b9e61bfdef6c0637c5c2a42cf0;l=16)) depending on how we execute this test. Can you change `0xa.......` to `0xf.......`?

Comment (alphan) [inline sw/device/tests/rstmgr_cpu_info_test.c:32]: I don't think this is the pattern we want. Please use either the `mmio` or `abs_mmio` libraries and replace these with an anonymous `enum`, i.e.:
```
enum {
   kIllegalAddr0 = 0xf.......,
   ...
};

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 10]
```diff
@@ -24,176 +23,253 @@ OTTF_DEFINE_TEST_CONFIG();
...
 
 /**
- * Reserve expected cpu dump area in flash
+ * When true, the exception handler will trigger another fault,
+ * causing a double fault,
+ * otherwise it triggers a software reset.
  */
-OT_SECTION(".non_volatile_scratch")
-dif_rstmgr_cpu_info_dump_segment_t exp_dump[kCpuDumpSize];
+static bool double_fault;
```
Comment (alphan) [inline sw/device/tests/rstmgr_cpu_info_test.c:60]: This is shared between the handler and the main program, thus should be marked as volatile.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 11]
```diff
@@ -24,176 +23,253 @@ OTTF_DEFINE_TEST_CONFIG();
...
+    kSingleFaultAddrCurrentPc[], kSingleFaultAddrNextPc[],
+    kDoubleFaultFirstAddrLower[], kDoubleFaultFirstAddrUpper[],
+    kDoubleFaultSecondAddrLower[], kDoubleFaultSecondAddrUpper[];
+
+// A handle to the reset manager.
+static dif_rstmgr_t rstmgr;
+
 /**
- * Cpu dump struct index
+ * This variable is used to ensure loads from an address aren't optimised out.
```
Comment (alphan) [inline sw/device/tests/rstmgr_cpu_info_test.c:51]: I think we need this to be volatile just because this is shared between the handler and the main program.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 12]
```diff
@@ -24,176 +23,253 @@ OTTF_DEFINE_TEST_CONFIG();
...
+ * @param mpec_upper The expected range upper bound of the last exception PC.
+ * @param last_data_addr The address of the last data access.
+ * @param curr_pc The current PC.
+ * @param next_pc The last PC.
+ */
+static void check_current_values(dif_rv_core_ibex_crash_dump_info_t *obs_dump,
+                                 uint32_t last_exception_addr,
+                                 uint32_t mpec_lower, uint32_t mpec_upper,
+                                 uint32_t last_data_addr, uint32_t curr_pc,
+                                 uint32_t next_pc) {
```
Comment (alphan) [inline sw/device/tests/rstmgr_cpu_info_test.c:125]: Please use a struct (or define a new one if you cannot use the definitions in `dif_rv_core_ibex`)  instead of defining 6 `uint32_t` parameters to help avoid mistakes. Call-sites then can become:
```
check_current_vaules(foo, (my_struct_t) {
    .last_exception_addr = some_val,
    .mpec_lower = another_val,
    ...
    });
```
Also applies to the function below.

**Review by alphan — COMMENTED** [review]
Thanks @HU90m, please see my comments.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 13]
```diff
@@ -24,176 +23,253 @@ OTTF_DEFINE_TEST_CONFIG();
...
+                           /*mpec_upper=*/(uint32_t)kDoubleFaultSecondAddrUpper,
+                           /*last_data_addr=*/(uint32_t)kIllegalAddr2,
+                           /*curr_pc=*/curr_pc,
+                           /*next_pc=*/next_pc);
+      check_previous_values(
+          &dump,
+          /*last_exception_addr=*/(uint32_t)kIllegalAddr1,
+          /*mpec_lower=*/(uint32_t)kDoubleFaultFirstAddrLower,
+          /*mpec_upper=*/(uint32_t)kDoubleFaultFirstAddrUpper);
+      break;
```
Comment (alphan) [inline sw/device/tests/rstmgr_cpu_info_test.c:269]: cannot comment on an already existing line but please consider `return`ing `false` by default at the end of the function and `return`in `true` here so that any new branches fail by default in the future. I'm not sure how critical it is to disable the timer though. I don't think our tests are expected to clean-up after themselves.

File: sw/device/lib/dif/dif_rv_core_ibex.h (hunk 4 above)
Comment (HU90m) [reply sw/device/lib/dif/dif_rv_core_ibex.h:204]: Yes that would be neater and...     done.

File: sw/device/tests/rstmgr_cpu_info_test.c (hunk 7 above)
Comment (HU90m) [reply sw/device/tests/rstmgr_cpu_info_test.c]: Yeah that would be nice. Changed it.

File: sw/device/tests/rstmgr_cpu_info_test.c (hunk 9 above)
Comment (HU90m) [reply sw/device/tests/rstmgr_cpu_info_test.c]: Good spot, thank you. All changed.

Comment (HU90m) [reply sw/device/tests/rstmgr_cpu_info_test.c]: Changed to this pattern.

File: sw/device/tests/rstmgr_cpu_info_test.c (hunk 13 above)
Comment (HU90m) [reply sw/device/tests/rstmgr_cpu_info_test.c]: That is a good point the default is not `return false`. I have left in the clean-up for the success case just in-case.

File: sw/device/tests/rstmgr_cpu_info_test.c (hunk 12 above)
Comment (HU90m) [reply sw/device/tests/rstmgr_cpu_info_test.c]: Yes, that is far nicer. Changed to this.

File: sw/device/tests/rstmgr_cpu_info_test.c (hunk 11 above)
Comment (HU90m) [reply sw/device/tests/rstmgr_cpu_info_test.c]: Because the `addr_val` isn't used after the assignments in the single fault case in the exception handler (line 93 in the latest commit), I thought the assignments would also be optimized out if it wasn't volatile.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 14]
```diff
@@ -22,181 +21,291 @@ OTTF_DEFINE_TEST_CONFIG();
...
-  kCpuDumpIdxCurrentExceptionPc = 1,
-  kCpuDumpIdxCurrentLastDataAddr = 2,
-  kCpuDumpIdxCurrentNextPc = 3,
-  kCpuDumpIdxCurrentPc = 4,
-  kCpuDumpIdxPreviousExceptionAddr = 5,
-  kCpuDumpIdxPreviousExceptionPc = 6,
-  kCpuDumpIdxPreviousValid = 7,
+  kIllegalAddr0 = 0xF0000000,
+  kIllegalAddr1 = 0xF0000001,
+  kIllegalAddr2 = 0xF0000002,
```
Comment (luismarques) [inline sw/device/tests/rstmgr_cpu_info_test.c:47]: It might not make a difference here, but often it's good to ensure that the separate addresses differ by at least the word size (or, conversely, also have cases where that's not the case), as that's often an edge case.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 15]
```diff
@@ -24,179 +23,186 @@ OTTF_DEFINE_TEST_CONFIG();
...
- *  After the double fault, the dut gets reset by the watch dog bite,
+ *  After the double fault, the DUT gets reset by the watchdog bite
  *  and the test collects / checks the cpu_info from the rstmgr.
  */
 
-// non existing address
-#define kIllegalAddr1 0x4041FFF0u
-#define kIllegalAddr2 0x40003618u
-#define kSkipComp 0x12345678u
 #define kCpuDumpSize 8
```
Comment (engdoreis) [inline sw/device/tests/rstmgr_cpu_info_test.c:41]: Nit. defines should be named with Captals, but in this case I would move this const to the enum below.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 16]
```diff
@@ -24,179 +23,186 @@ OTTF_DEFINE_TEST_CONFIG();
...
+  CHECK_DIF_OK(
+      dif_rv_core_ibex_parse_crash_dump(ibex, dump, size_read, &output));
+  return output;
+}
 
-  read_error();
+typedef struct rstmgr_cpu_info_test_exp_state {
+  uint32_t mtval;   // The last exception address.
+  uint32_t mpec_l;  // The last exception PC lower bound
+  uint32_t mpec_u;  // The last exception PC upper bound
```
Comment (engdoreis) [inline sw/device/tests/rstmgr_cpu_info_test.c]: Nit. A full stop is missing in the two lines above.

File: sw/device/tests/rstmgr_cpu_info_test.c [hunk 17]
```diff
@@ -21,50 +21,95 @@ OTTF_DEFINE_TEST_CONFIG();
...
+extern const char kSingleFaultAddrNextPc[];
 extern const char kDoubleFaultFirstAddrLower[];
 extern const char kDoubleFaultFirstAddrUpper[];
 extern const char kDoubleFaultSecondAddrLower[];
 extern const char kDoubleFaultSecondAddrUpper[];
 
+/// A handle to the reset manager.
+static dif_rstmgr_t rstmgr;
+
+/// This variable is used to ensure loads from an address aren't optimised out.
```
Comment (engdoreis) [inline sw/device/tests/rstmgr_cpu_info_test.c:67]:
```Suggestion
// This variable is used to ensure loads from an address aren't optimised out.
```
For consistency.
The same for the line 64.

**Review by engdoreis — APPROVED** [review]

File: sw/device/tests/rstmgr_cpu_info_test.c (hunk 14 above)
Comment (HU90m) [reply sw/device/tests/rstmgr_cpu_info_test.c]: They are only dereferenced into `uint8_t`, but I use `mmio_region_read32` so I think it's a fair point.
I have separated them into 32bit aligned addresses, also made `addr_val` a `uint32_t`.

File: sw/device/tests/rstmgr_cpu_info_test.c (hunk 15 above)
Comment (HU90m) [reply sw/device/tests/rstmgr_cpu_info_test.c]: Done

**Review by alphan — COMMENTED** [review]
LGTM, just a nit.
